TrackletAlgorithm/TrackBuilder.h

```diff
@@ -0,0 +1,276 @@
+#ifndef TrackletAlgorithm_TrackBuilder_h
+#define TrackletAlgorithm_TrackBuilder_h
+
+#include "TrackletParameterMemory.h"
+#include "FullMatchMemory.h"
+#include "TrackFitMemory.h"
+
+static const unsigned short kNBitsTBBuffer = 1;
+static const unsigned short kMinNMatches = 2;
+static const unsigned short kInvalidTrackletID = 0x3FFF;
+
+static const unsigned short kNBitsTCID = FullMatchBase<BARREL>::kFMTCIDSize;
+static const unsigned short kNBitsTrackletID = kNBitsTCID + kNBits_MemAddr;
+
+typedef ap_uint<kNBitsTCID> TCIDType;
+typedef ap_uint<kNBits_MemAddr> IndexType;
+typedef ap_uint<kNBitsTrackletID> TrackletIDType;
+
+// Slim data type used to store full match information in the circular buffers.
+struct MyStub {
+  public:
+    MyStub() {}
+    MyStub(const IndexType &index, const TrackletIDType &id) {
+      setIndex(index);
+      setID(id);
+    }
+    const IndexType index() const { return data_.range(kNBits_MemAddr - 1, 0); }
+    const TrackletIDType id() const { return data_.range(kNBits_MemAddr + kNBitsTrackletID - 1, kNBits_MemAddr); }
+    void setIndex(const IndexType &index) { data_.range(kNBits_MemAddr - 1, 0) = index; }
+    void setID(const TrackletIDType &id) { data_.range(kNBits_MemAddr + kNBitsTrackletID - 1, kNBits_MemAddr) = id; }
+  private:
+    ap_uint<kNBitsTrackletID + kNBits_MemAddr> data_;
+};
+
+// Function to retrieve a full match and store the associated index and
+// tracklet ID in one element of a circular buffer.
+template<regionType RegionType> void
+getFM(const BXType bx, const FullMatchMemory<RegionType> &fullMatches, const unsigned short i, MyStub &fm)
+{
+  if (i < fullMatches.getEntries(bx)) {
+    const auto &full_match = fullMatches.read_mem(bx, i);
+    fm.setIndex(i);
+    fm.setID(full_match.getTrackletID());
+  }
+  else {
+    // If there are no more full matches, set the index and ID to dummy values.
+    fm.setIndex(0);
+    fm.setID(kInvalidTrackletID);
+  }
+}
+
+// TrackBuilder top template function
+// !!! CURRENTLY ONLY TESTED FOR L1L2 !!!
+template<unsigned NTPar, unsigned NFMBarrel, unsigned NFMDisk>
+void TrackBuilder(
+    const BXType bx,
+    const TrackletParameterMemory trackletParameters[NTPar],
+    const FullMatchMemory<BARREL> barrelFullMatches[NFMBarrel],
+    const FullMatchMemory<DISK> diskFullMatches[NFMDisk],
+    BXType &bx_o,
+    TrackFit::TrackWord trackWord[kMaxProc],
+    TrackFit::BarrelStubWord barrelStubWords[4][kMaxProc],
```

Comment:
Eliminate magic number "4". (Perhaps replace with const in TrackFitMemory.h?)

Reply:
Done.